File: sc_project/events/quotation.py
import frappe

@frappe.whitelist()
def make_new_bom(quotation, title, bom_qty, quotation_item,existing=None):
    if existing:
        bom = frappe.copy_doc(frappe.get_doc("Estimation BOM",existing))
        bom.insert()
        return bom.name
    else:
        bom = frappe.get_doc({
            "doctype":"Estimation BOM",
            "quotation":quotation,
            "title":title,
            "bom_qty":bom_qty,
            "quotation_item":quotation_item,
            }).insert()
        return bom.name

def delete_unlinked_boms(doc,method=None):
    unlinked_docs = frappe.get_all("Estimation BOM",{"quotation":doc.name})
    for bom in unlinked_docs:
        if not bom.name in [row.custom_estimation_bom for row in doc.items ]:
            frappe.delete_doc("Estimation BOM",bom.name)
            frappe.msgprint(f"BOM {bom.name} Deleted!")

def quotation_duplicate(doc,method=None):
    for item in doc.items:
        if item.custom_estimation_bom:
            new_estimation_bom = frappe.copy_doc(frappe.get_doc("Estimation BOM",item.custom_estimation_bom))
            new_estimation_bom.quotation = doc.name
            new_estimation_bom.quotation_item = item.name
            new_estimation_bom.insert()
            item.db_set("custom_estimation_bom",new_estimation_bom.name)

File: sc_project/public/js/quotation.js
frappe.ui.form.on('Quotation', {
    refresh(frm) {
    }
})
frappe.ui.form.on('Quotation Item', {
    custom_open_bom(frm, cdt, cdn) {
        let row = locals[cdt][cdn]
        makeBOM(frm, row)
    }
})

let makeBOM = (frm, row) => {
    if (row.custom_estimation_bom) {
        frappe.set_route("Form", "Estimation BOM", row.custom_estimation_bom)
    }
    else {
        frappe.prompt({ label: "Copy From BOM", fieldtype: "Link", options: "Estimation BOM", fieldname: 'bom' }, v => {
            frappe.call("sc_project.events.quotation.make_new_bom",
                {
                    quotation: frm.doc.name,
                    title: row.item_name,
                    bom_qty: row.qty,
                    quotation_item: row.name,
                    existing: v.bom
                }).then(r => {
                    frappe.model.set_value(row.doctype, row.name, "custom_estimation_bom", r.message)
                    frm.save()
                    frappe.set_route("Form", "Estimation BOM", r.message)
                })
        })
    }
}